store/sqlstore/reaction_store.go

```diff
@@ -167,23 +178,80 @@ func (s *SqlReactionStore) PermanentDeleteBatch(endTime int64, limit int64) (int
 	return rowsAffected, nil
 }
 
-func saveReactionAndUpdatePost(transaction *gorp.Transaction, reaction *model.Reaction) error {
-	if err := transaction.Insert(reaction); err != nil {
-		return err
+func (s *SqlReactionStore) saveReactionAndUpdatePost(transaction *gorp.Transaction, reaction *model.Reaction) error {
+	params := map[string]interface{}{
+		"UserId":    reaction.UserId,
+		"PostId":    reaction.PostId,
+		"EmojiName": reaction.EmojiName,
+		"CreateAt":  reaction.CreateAt,
+		"UpdateAt":  reaction.UpdateAt,
 	}
 
+	if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		if _, err := transaction.Exec(
+			`INSERT INTO
+				Reactions
+				(UserId, PostId, EmojiName, CreateAt, UpdateAt, DeleteAt)
+			VALUES
+				(:UserId, :PostId, :EmojiName, :CreateAt, :UpdateAt, 0)
+			ON DUPLICATE KEY UPDATE
+				UpdateAt = :UpdateAt, DeleteAt = 0`, params); err != nil {
+			return err
+		}
+	} else if s.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		// postgres has no way to upsert values until version 9.5 and trying inserting and then updating causes transactions to abort
+		count, err := transaction.SelectInt(
+			`SELECT
```

Comment:
I'd prefer to use an `ON CONFLICT` here. It doesn't seem like a lot of work. Otherwise, we can simply use a CTE. There is no need to have 2 separate queries.

Reply:
Now using ON CONFLICT  https://github.com/mattermost/mattermost-server/pull/16654/commits/474429d278b1475ed9b7ec1690b3611b291b0360